Update resize op

daquexian opened the pull request:
1. Add "cubic" interpolation mode, and clarify the "linear", "bilinear" and "trilinear" (https://github.com/onnx/onnx/issues/1774)
2. Update the tests: The current tests of resize linear follows the old behavior of TF, which is changed now. Related issue: https://github.com/tensorflow/tensorflow/issues/6720, https://github.com/onnx/onnx/issues/2070
3. Add "coordinate_transformation_mode" attribute including "half_pixel", "align_corners", "asymmetric", "tf_crop_and_resize" and so on
4. Add a new input 'sizes' as requested in #2062.
5. Add two new attrs 'cubic_coeff_a' and 'exclude_outside' for the compatibility with TensorFlow. There is no standard implementation of cubic interpolation. [TensorFlow](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/resize_bicubic_op.cc#L60) (and also [matlab](https://theailearner.com/2018/12/29/image-processing-bicubic-interpolation/)) sets a=-0.5 (In the legacy version of TF, a=-0.75) and [exclude_outside=True](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/resize_bicubic_op.cc#L95), while PyTorch (and OpenCV) sets a=-0.75 and exclude_outside=False

Both cubic interpolation mode and align_corners attribute are supported by [tensorflow](https://www.tensorflow.org/api_docs/python/tf/image/resize_bicubic), [pytorch](https://pytorch.org/docs/stable/nn.html#torch.nn.functional.interpolate) and so on.

Note: 
* Some frameworks set `output_dimension` to `round(input_dimension * scale)` instead of `floor(input_dimension * scale)`, some frameworks re-assign the `scale` to `output_dimension / input_dimension`. I think it is not good to add two extra attributes in resize op for these cases. This op has many attributes already, and the converters in frameworks have the ability and responsibility to make the `scale` compatible with the behavior of this op.

* This proposal is fully compatible with TensorFlow and PyTorch. Please refer to [the checking script](https://gist.github.com/daquexian/4bab98a7e706ad95616567433c1234b8 ) for the attribute values

daquexian commented:
About the existing example/test:

It is following the behavior of tensorflow, which has been considered as a bug and fixed in TensorFlow nightly (https://github.com/tensorflow/tensorflow/issues/6720)

```python
import tensorflow as tf
import numpy as np

print(tf.__version__)
a=np.array([[1,2],[3,4]])
input=tf.constant(a)
input4d=tf.reshape(input, [1,2,2,1])
resize=tf.compat.v2.image.resize(input4d,[4,4])
sess=tf.Session()
r1 = sess.run(resize)
print(r1)
```
gives
```python
1.14.1-dev20190526
[[[[1.  ]
   [1.25]
   [1.75]
   [2.  ]]

  [[1.5 ]
   [1.75]
   [2.25]
   [2.5 ]]

  [[2.5 ]
   [2.75]
   [3.25]
   [3.5 ]]

  [[3.  ]
   [3.25]
   [3.75]
   [4.  ]]]]
```

daquexian commented:
I have updated the tests :)

linkerzhang commented:
@postrational

linkerzhang commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1355,9 +1355,24 @@ ONNX_OPERATOR_SET_SCHEMA(
     OpSchema()
```
please bump the op version. You may refer to https://github.com/onnx/onnx/blob/master/docs/Versioning.md#operator-versioning to check onnx versioning design.

wschin commented on onnx/backend/test/case/node/resize.py:
```diff
@@ -104,8 +132,170 @@ def export_downsample_linear():  # type: () -> None
+        scales = np.array([1.0, 1.0, 2.0, 2.0], dtype=np.float32)
+
+        output = np.array([[[
+            [ 0.47265625,  0.76953125,  1.24609375,  1.87500000,  2.28125000,
```
Where do these numbers come from? It's not straightforward to write code to implement this behavior. Would you consider adding more `math` details and a lightweight numpy implementation here?

daquexian replied:
@wschin 

These numbers come from pytorch and opencv. The result of opencv is the same as pytorch with align_corners=False.

I have implemented linear and cubic interpolation [here](https://gist.github.com/daquexian/18b3fda2c7d2a6ad236a0a5ded4fe845) :)

wschin replied:
Could you please have a Python version of your implementation (only depending on numpy) and compute the test outputs using that Python implementation? This operator is important and also very sophisticated, so a reference implementation is desired. Please take a look at #2066 for more reasons.

daquexian replied:
All right, thanks for your suggestion :) I'll do it soon.

linkerzhang commented:
please also fix the ci failure due to flake8 errors.

like, 
+flake8
./onnx/backend/test/case/node/resize.py:176:14: E201 whitespace after '['
./onnx/backend/test/case/node/resize.py:176:17: E241 multiple spaces after ','
./onnx/backend/test/case/node/resize.py:176:21: E241 multiple spaces after ','
.....

daquexian commented:
@linkerzhang I have upgraded the op version and fixed flake8 errors. :)

daquexian commented:
I have figured out that I should add the upgraded operator in `onnx/defs/operator_sets.h`. Now the docs look right.

daquexian commented:
I will resolve the conflict of Changelog.md soon.

I added a new input 'sizes' as requested in https://github.com/onnx/onnx/issues/2062. 

Furthermore, I added two new attrs, 'cubic_coeff_a' and 'exclude_outside', for the compatibility with TensorFlow. TensorFlow sets [A=-0.5](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/resize_bicubic_op.cc#L60) (In its legacy version, A=-0.75) and [exclude_outside=True](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/resize_bicubic_op.cc#L95), while PyTorch (and OpenCV) sets A=-0.75 and exclude_outside=False

daquexian commented:
@linkerzhang @wschin I have added the implementation in the tests. It is ready to review. I'll implement the shape inference after https://github.com/onnx/onnx/pull/2085 is merged.

lara-hdr commented:
Is this ready? We are also waiting for this PR, @ebarsoum @houseroad

daquexian commented:
@lara-hdr I'll update the shape inference soon, and then this PR will be ready to review :)

daquexian commented:
@ebarsoum @linkerzhang @houseroad @wschin I have updated shape inference. This PR is ready to review :)

daquexian commented:
@linkerzhang I have checked the proposed RoiCropAndResize op. I found that it is actually unable to compose it by Slice and Resize ops. The coordinates of rois can be non-integer, however, the Slice op can only generate an output tensor with integer height and width. Maybe we can combine RoiCrpAndResize op with Resize op by adding "crop" and "extrapolation_value" attribute?

jiafatom commented:
> @linkerzhang I have checked the proposed RoiCropAndResize op. I found that it is actually unable to compose it by Slice and Resize ops. The coordinates of rois can be non-integer, however, the Slice op can only generate an output tensor with integer height and width. Maybe we can combine RoiCrpAndResize op with Resize op by adding "crop" and "extrapolation_value" attribute?
> 
> @jiafatom @ebarsoum @BowenBao What's your opinion? :)
What do you mean "combine RoiCrpAndResize op with Resize op by adding "crop" and "extrapolation_value" attribute?" 
RoiCropAndResize op has extrapolation_value attribute. Where do you want to add "crop" and "extrapolation_value"? To Resize op?

daquexian commented:
> > @linkerzhang I have checked the proposed RoiCropAndResize op. I found that it is actually unable to compose it by Slice and Resize ops. The coordinates of rois can be non-integer, however, the Slice op can only generate an output tensor with integer height and width. Maybe we can combine RoiCrpAndResize op with Resize op by adding "crop" and "extrapolation_value" attribute?
> > @jiafatom @ebarsoum @BowenBao What's your opinion? :)
> 
> What do you mean "combine RoiCrpAndResize op with Resize op by adding "crop" and "extrapolation_value" attribute?"
> RoiCropAndResize op has extrapolation_value attribute. Where do you want to add "crop" and "extrapolation_value"? To Resize op?

Yes, added crop and extrapolation_value attributes to resize op.

jiafatom commented:
Resize is a special case of RoiCropAndResize, where the roi for Resize is [0, 0, 1, 1] (normalized coordinates). On the other hand, the current RoiCropAndResize is used only for 2d case, there is no use case to extend it to n-d.

So I feel it not good to add crop and extrapolation_value on resize op, because these two are only used in RoiCropAndResize.

daquexian commented:
@jiafatom thanks for your reply. Do you have any ideas on how to define roicropandresize as a function? I think the problem is the non-integer rois (as said in https://github.com/onnx/onnx/pull/2057#issuecomment-510722131)

jiafatom commented:
To define roicropandresize as a function, we need enhance resize op to make it like a crop_and_resize for a single ROI. Then like you said, we need add one input as crop, the other attributes as extrapolation_value. This way we make resize op as a general roicropandresize. Actually I have concerns about this because there is no use case for n-d roicropandresize other than 2d, it is over-killing the problem.

Another thing: Why non-integer roi is a problem? The input roi range is from 0 to 1 because it is normalized. if it is [0, 0, 1, 1], then it is actually original resize. No matter it is an integer or not, we need find the corresponding pixel in the input tensor (which is also a floating number), and then apply specific modes such as bilinear, nearest (it will use the integer coordinates in the original input).

daquexian commented:
> Why non-integer roi is a problem? The input roi range is from 0 to 1 because it is normalized. if it is [0, 0, 1, 1], then it is actually original resize. No matter it is an integer or not, we need find the corresponding pixel in the input tensor (which is also a floating number), and then apply specific modes such as bilinear, nearest (it will use the integer coordinates in the original input).

What I mean is, we cannot define cropandresize as a function based on slice op and the existing resize op as https://github.com/onnx/onnx/pull/2142#issuecomment-507898382 since the output of slice has integer size but the roi has float size. As a result, we have to expand resize op by adding a bulit-in crop attribute.

daquexian commented:
> Actually I have concerns about this because there is no use case for n-d roicropandresize other than 2d, it is over-killing the problem.
@jiafatom I think it is necessary to add this "crop" input. There is no other way to handle the float roi coordinates. Please correct me if I am wrong.

jiafatom commented:
> @jiafatom I think it is necessary to add this "crop" input. There is no other way to handle the float roi coordinates. Please correct me if I am wrong.

I agree that we need add "crop" input to support roi coordinates :)

BowenBao commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1358,35 +1358,75 @@ ONNX_OPERATOR_SET_SCHEMA(
+static const char* Resize_ver11_doc = R"DOC(
 Resize the input tensor.
 Each dimension value of the output tensor is:
   output_dimension = floor(input_dimension * scale).
```
add "if input 'sizes' is not specified"

daquexian replied:
Thanks! I'll edit them soon.

daquexian replied:
Thanks! It is added.

BowenBao commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1358,35 +1358,75 @@ ONNX_OPERATOR_SET_SCHEMA(
             "mode",
-            "Two interpolation modes: nearest (default), and linear (including bilinear, trilinear, etc)",
+            "Three interpolation modes: nearest (default), linear and cubic. "
+            "The \"linear\" mode includes linear interpolation for 1D tensor and N-D linear interpolation for N-D tensor (For example, trilinear interpolation for 2D tensor). "
```
> "The \"linear\" mode includes linear interpolation for 1D tensor and N-D linear interpolation for N-D tensor (For example, trilinear interpolation for 2D tensor). "

should it be "_bilinear_ interpolation for 2D tensor" ?

daquexian replied:
Thanks! It is fixed.

BowenBao commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1358,35 +1358,75 @@ ONNX_OPERATOR_SET_SCHEMA(
-            "Two interpolation modes: nearest (default), and linear (including bilinear, trilinear, etc)",
+            "Three interpolation modes: nearest (default), linear and cubic. "
+            "The \"linear\" mode includes linear interpolation for 1D tensor and N-D linear interpolation for N-D tensor (For example, trilinear interpolation for 2D tensor). "
+            "The \"cubic\" mode includes cubic interpolation for 1D tensor and N-D cubic interpolation for N-D tensor (For example, tricubic interpolation for 2D tensor)."
```
same above, bicubic?

daquexian replied:
Thanks! It is fixed.

winnietsang commented:
@daquexian which Tensorflow operator will create the same result as your unit test. There are tf.image.resize_images and tf.compat.v2.image.resize.
tf.image.resize_images has align_corners as one of the args but tf.compat.v2.image.resize don't.
In your earlier comment you mention that you have try this on Tensorflow, may you please share with me which tf operator are you using? Thanks

daquexian commented:
> @daquexian which Tensorflow operator will create the same result as your unit test. There are tf.image.resize_images and tf.compat.v2.image.resize.
> tf.image.resize_images has align_corners as one of the args but tf.compat.v2.image.resize don't.
> In your earlier comment you mention that you have try this on Tensorflow, may you please share with me which tf operator are you using? Thanks

I tested with tf.compat.v2.image.resize. I'm not familiar with tensorflow. It seems that the result of tf.image.resize_images and tf.compat.v2.image.resize are different.

spandantiwari commented:
@daquexian - could you please confirm if all the feedback has been addressed. This is an important change and we  are waiting for this PR to merge in ONNX.

daquexian commented:
@spandantiwari I'll update this PR soon. Thanks :)

spandantiwari commented:
@daquexian - thanks a lot!

daquexian commented:
@winnietsang tf.v1 behavior with half_pixel_centers=False is supported by the new 'tf_legacy_scaler` attribute

daquexian commented:
@jiafatom I have updated this PR and added input "roi" and attr "extrapolation_value". I'll check whether the output is consistent with your PR tomorrow.

jiafatom commented on onnx/test/shape_inference_test.py:
```diff
@@ -305,7 +305,19 @@ def test_expand_raw_data(self):  # type: () -> None
+        graph = self._make_graph(
+            [('x', TensorProto.INT32, (2, 4, 3, 5)),
+             ('scales', TensorProto.FLOAT, (4,)),
+             ('sizes', TensorProto.INT64, (4,))],
```
add roi here?

daquexian replied:
Added. Thanks!

jiafatom commented on onnx/backend/test/case/node/resize.py:
```diff
@@ -8,12 +8,221 @@
+    return np.array([1])
 
 
 class Resize(Base):
```
add roi in all the test case?

daquexian replied:
Thanks! Will add.

daquexian replied:
Added. Thanks!

jiafatom commented on docs/Operators.md:
```diff
@@ -11663,43 +11663,161 @@ for test_name, shape in test_cases.items():
 <dd>N-D tensor</dd>
-<dt><tt>scales</tt> : tensor(float)</dt>
-<dd>The scale array along each dimension. It takes value greater than 0. If it's less than 1, it's sampling down, otherwise, it's upsampling. The number of elements of 'scales' should be the same as the rank of input 'X'.</dd>
+<dt><tt>scales</tt> : T2</dt>
```
roi missing?

daquexian replied:
Thanks! I have not re-generate the doc yet.

daquexian replied:
Done. Thanks!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/onnx/onnx?pullRequest=2057) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/onnx/onnx?pullRequest=2057) before we can accept your contribution.<br/>**0** out of **3** committers have signed the CLA.<br/><br/>:x: daquexian<br/>:x: linkerzhang<br/>:x: houseroad<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/onnx/onnx?pullRequest=2057) it.</sub>

daquexian commented:
@jiafatom I have updated this PR. The output of my implementation is the same as yours. :)

daquexian commented:
@ebarsoum @linkerzhang @houseroad @wschin @jiafatom @BowenBao @spandantiwari It is ready to review :)

winnietsang commented:
@daquexian I'm in the process to verify all of your test cases using Tensorflow ops, I will update you shortly once I have some preliminary result.

daquexian commented:
> @daquexian I'm in the process to verify all of your test cases using Tensorflow ops, I will update you shortly once I have some preliminary result.

Please set scaler="tf_legacy" for the case half_pixel_centers=False, and set scaler="half_pixel", cubic_coeff_a=-0.5, exclude_outside=True for the case half_pixel_center=True

winnietsang commented:
> > @daquexian I'm in the process to verify all of your test cases using Tensorflow ops, I will update you shortly once I have some preliminary result.
> 
> Thanks! Please set scaler="tf_legacy" for the case half_pixel_centers=align_corners=False, and set scaler="half_pixel", cubic_coeff_a=-0.5, exclude_outside=True for the case half_pixel_center=True

@daquexian I try the 4 testcases for mode=nearest and here are my findings:

1. upsample_scales_nearest: able to get the same values as yours with 
tf.compat.v2.image.resize, 
tf.compat.v1.image.resize_nearest_neighbor(half_pixel_centers=False) and tf.compat.v1.image.resize_nearest_neighbor(half_pixel_centers=True)
 
2. downsample_scales_nearest: can only get the same values as yours with     tf.compat.v1.image.resize_nearest_neighbor(half_pixel_centers=False)
but got different values with 
tf.compat.v2.image.resize and    
tf.compat.v1.image.resize_nearest_neighbor(half_pixel_centers=True)

3. upsample_size_nearest: can only get the same values as yours with 
tf.compat.v1.image.resize_nearest_neighbor(half_pixel_centers=False)  
but got different values with 
tf.compat.v2.image.resize and    
tf.compat.v1.image.resize_nearest_neighbor(half_pixel_centers=True)

4. downsample_size_nearest: got different values with 
tf.compat.v2.image.resize, 
tf.compat.v1.image.resize_nearest_neighbor(half_pixel_centers=False) and tf.compat.v1.image.resize_nearest_neighbor(half_pixel_centers=True)

Based on above results, I think the implementations of half_pixel_centers in onnx and tensorflow are different. Can you possibly look into this?

I also changed the initial values for case 1, scale = [1.0, 1.0, 2.5, 3.0], and saw the results not match with the expected values in tf.compat.v2.image.resize or tf.compat.v1.image.resize_nearest_neighbor(half_pixel_centers=True). So this seems to again indicate the half_pixel_centers has significant impact to the expected results.

daquexian commented:
> > > @daquexian I'm in the process to verify all of your test cases using Tensorflow ops, I will update you shortly once I have some preliminary result.
> > 
> > 
> > Thanks! Please set scaler="tf_legacy" for the case half_pixel_centers=align_corners=False, and set scaler="half_pixel", cubic_coeff_a=-0.5, exclude_outside=True for the case half_pixel_center=True
> 
> @daquexian I try the 4 testcases for mode=nearest and here are my findings:
> 
> 1. upsample_scales_nearest: able to get the same values as yours with
>    tf.compat.v2.image.resize,
>    tf.compat.v1.image.resize_nearest_neighbor(half_pixel_centers=False) and tf.compat.v1.image.resize_nearest_neighbor(half_pixel_centers=True)
> 2. downsample_scales_nearest: can only get the same values as yours with     tf.compat.v1.image.resize_nearest_neighbor(half_pixel_centers=False)
>    but got different values with
>    tf.compat.v2.image.resize and
>    tf.compat.v1.image.resize_nearest_neighbor(half_pixel_centers=True)
> 3. upsample_size_nearest: can only get the same values as yours with
>    tf.compat.v1.image.resize_nearest_neighbor(half_pixel_centers=False)
>    but got different values with
>    tf.compat.v2.image.resize and
>    tf.compat.v1.image.resize_nearest_neighbor(half_pixel_centers=True)
> 4. downsample_size_nearest: got different values with
>    tf.compat.v2.image.resize,
>    tf.compat.v1.image.resize_nearest_neighbor(half_pixel_centers=False) and tf.compat.v1.image.resize_nearest_neighbor(half_pixel_centers=True)
> 
> Based on above results, I think the implementations of half_pixel_centers in onnx and tensorflow are different. Can you possibly look into this?
> 
> I also changed the initial values for case 1, scale = [1.0, 1.0, 2.5, 3.0], and saw the results not match with the expected values in tf.compat.v2.image.resize or tf.compat.v1.image.resize_nearest_neighbor(half_pixel_centers=True). So this seems to again indicate the half_pixel_centers has significant impact to the expected results.

Thanks! I found there are some differences in the implementation of nearest resize. I'll add a new attribute for this case soon.

daquexian commented:
@winnietsang I have checked the result of this PR vs tensorflow on linear and cubic interpolation.

Here is the full code:

https://gist.github.com/daquexian/4bab98a7e706ad95616567433c1234b8

 The resize implementation of tf.v2 are the same with those of tf.v1 with half_pixel_centers=True, please check out [tf.v1](https://github.com/tensorflow/tensorflow/blob/v1.14.0/tensorflow/python/ops/image_ops_impl.py#L1102) and [tf.v2](https://github.com/tensorflow/tensorflow/blob/v1.14.0/tensorflow/python/ops/image_ops_impl.py#L1186) to validate it.

winnietsang commented:
@daquexian I try to verify the testcase for linear and cubic today, I got a lot of cases that the expected values doesn't match with the values returned from Tensorflow. 
Then I switch to look at the test code you refer https://gist.github.com/daquexian/4bab98a7e706ad95616567433c1234b8 yesterday. I try to run downsample_scales_linear in that code by changing h=1, w=2 and resize='linear' and I got matching result from expected and tensorflow values. I also try to run upsample_sizes_cubic and downsample_sizes_cubic cases in your test code, both get matching values. 
Then I go back to the onnx unit test for resize in this PR, I notice that not all cubic testcase with scaler=half_pixel have the exclude_outside attribute set. And the ﻿coeffs value are different in the onnx unit test and your test code. May you please look into this and validate the unit test in this PR match with the test code you provide yesterday? Thanks

daquexian commented:
> @daquexian I try to verify the testcase for linear and cubic today, I got a lot of cases that the expected values doesn't match with the values returned from Tensorflow.
> Then I switch to look at the test code you refer https://gist.github.com/daquexian/4bab98a7e706ad95616567433c1234b8 yesterday. I try to run downsample_scales_linear in that code by changing h=1, w=2 and resize='linear' and I got matching result from expected and tensorflow values. I also try to run upsample_sizes_cubic and downsample_sizes_cubic cases in your test code, both get matching values.
> Then I go back to the onnx unit test for resize in this PR, I notice that not all cubic testcase with scaler=half_pixel have the exclude_outside attribute set. And the ﻿coeffs value are different in the onnx unit test and your test code. May you please look into this and validate the unit test in this PR match with the test code you provide yesterday? Thanks

Sorry for not making myself clear. 

There are many frameworks and they are different implementations on resize op. To be compatible with every framework, some attributes like "exclude_outside" are introduced. TensorFlow expects exclude_outside to be True while PyTorch expects it to be false (when scaler='half_pixel'). I have to cover all these cases in tests. So only some of the tests are for TensorFlow, while others are for other frameworks like PyTorch. As a result, when testing this PR against TensorFlow, not every test can be used directly, one should always set attributes according to what TensorFlow expects.

daquexian commented:
@winnietsang I have updated this PR and also my gist https://gist.github.com/daquexian/4bab98a7e706ad95616567433c1234b8, now it is fully compatible with TensorFlow.

winnietsang commented:
> @winnietsang I have updated this PR and also my gist https://gist.github.com/daquexian/4bab98a7e706ad95616567433c1234b8, now it is fully compatible with TensorFlow (and also PyTorch)

@daquexian, thanks for the explanation of the unit test.

daquexian commented:
@winnietsang Thanks for your careful check! I have updated them

daquexian commented:
@houseroad @linkerzhang @wschin Is this PR ok to merge? I think I have addressed all comments :)

linkerzhang commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1361,41 +1361,103 @@ ONNX_OPERATOR_SET_SCHEMA(
-            "tensor(float)")
-        .Output(0, "Y", "N-D tensor after resizing", "T")
+            " be the same as the rank of input 'X'. It will be ignored if 'sizes' is specified.",
+            "T3")
```
put "tensor(float)" here directly.

daquexian replied:
Done :)

linkerzhang commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1361,41 +1361,103 @@ ONNX_OPERATOR_SET_SCHEMA(
+            "sizes",
+            "The size of the output tensor. The number of elements of 'sizes' should be the same as the"
+            " rank of input 'X'. 'scales' will be ignored if 'sizes' is specified.",
+            "T4",
```
put "tensor(int64)" here directly.

daquexian replied:
Done :)

daquexian commented:
@linkerzhang Thanks. It has been updated

wschin commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1395,40 +1395,94 @@ ONNX_OPERATOR_SET_SCHEMA(
-        .Input(0, "X", "N-D tensor", "T")
+        .Attr("cubic_coeff_a",
+            "The coefficient 'a' used in cubic interpolation. Two common choice are -0.5 (in TensorFlow) and -0.75"
+            " (in PyTorch). Check out https://ieeexplore.ieee.org/document/1163711 for the details.",
```
Do you mean `a` in the Equation (4) in the reference paper?

daquexian replied:
Yes :)

wschin replied:
Sounds good. Let's me more specific,
```suggestion
            " (in PyTorch). Check out Equation (4) in https://ieeexplore.ieee.org/document/1163711 for the details.",
```

wschin commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1395,40 +1395,94 @@ ONNX_OPERATOR_SET_SCHEMA(
-            "Two interpolation modes: nearest (default), and linear (including bilinear, trilinear, etc)",
+            "Three interpolation modes: nearest (default), linear and cubic. "
+            "The \"linear\" mode includes linear interpolation for 1D tensor and N-D linear interpolation for N-D tensor (For example, bilinear interpolation for 2D tensor). "
+            "The \"cubic\" mode includes cubic interpolation for 1D tensor and N-D cubic interpolation for N-D tensor (For example, bicubic interpolation for 2D tensor)."
```
I can't figure out how to implement this operator after reading this spec. Are you using this equation (copied from your reference paper)?
![image](https://user-images.githubusercontent.com/3524474/62344992-3eef8980-b4a5-11e9-8ac7-fff5cb24ca52.png)

daquexian replied:
N-D (2-D is a special case where N=2) interpolation is implemented by performing 1-D interpolation recursively. For N=2, assume we want to get the value at `(1.6, 2.8)` in resized 2-D HxW tensor, we firstly calculate H values at `(1.6)` by performing 1-D interpolation on H rows (a.k.a. H 1-D tensors) respectively, and then calculate the value at `(2.8)` of the generate line (a.k.a. 1-D tensor) consisting of H above values. Similarly, 3-D interpolation is 2-D interpolation on every layer (a.k.a. every 2-D tensor) and 1-D interpolation on the generated line (a.k.a. 1-D tensor).

wschin commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1395,40 +1395,94 @@ ONNX_OPERATOR_SET_SCHEMA(
+            AttributeProto::FLOAT,
+            static_cast<float>(-0.75))
+        .Attr("exclude_outside",
+            "If set to 1, the weight of sampling locations outside the tensor will be set to 0"
```
What is `weight`? Without a equation like 
![image](https://user-images.githubusercontent.com/3524474/62345182-1320d380-b4a6-11e9-84f4-6ea5d0db0768.png)
, it's not easy to explain those attributes in a way that the reader can implement this operator.

daquexian replied:
I have added some description in 1c87d37. Is it enough? Thanks!

wschin replied:
Looks good to me. Thank you.

wschin commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1395,40 +1395,94 @@ ONNX_OPERATOR_SET_SCHEMA(
+            static_cast<int64_t>(0))
+        .Attr("scaler",
+            "The scaler used in interpolation. "
+            "For each dimension, denote x_resized as the coordinate in the resized tensor, x_original as the corresponding coordinate in the original tensor, length_original as the length of the original tensor in the specific dimension, length_resized as the length of the resized tensor in the specific dimension, roi_x = (start, end) of the corresponding dimension in input \"roi\", scale = lengt ...
```
length_original as the length of the original tensor in the `specific dimension`? If it's for a specific dimension, would it be clearer to use length_original_0 for axis 0 and so on?

daquexian replied:
Given the context that N-D interpolation is indeed a combination of 1-D interpolation, symbols here can cover all axes. I'll try to make this paragraph more clear.

daquexian replied:
I have added some description in ff74643

wschin commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1395,40 +1395,94 @@ ONNX_OPERATOR_SET_SCHEMA(
+            "For each dimension, denote x_resized as the coordinate in the resized tensor, x_original as the corresponding coordinate in the original tensor, length_original as the length of the original tensor in the specific dimension, length_resized as the length of the resized tensor in the specific dimension, roi_x = (start, end) of the corresponding dimension in input \"roi\", scale = lengt ...
+            "if scaler is \"half_pixel\", "
+            "x_original = (x_resized + 0.5) / scale - 0.5, "
+            "if scaler is \"pytorch_half_pixel\", "
```
Could we break this long sentence into smaller ones?

daquexian replied:
Thanks! I will do it

wschin commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1395,40 +1395,94 @@ ONNX_OPERATOR_SET_SCHEMA(
+            " and the weight will be renormalized so that their sum is 1.0. The default value is 0.",
+            AttributeProto::INT,
+            static_cast<int64_t>(0))
+        .Attr("scaler",
```
```suggestion
        .Attr("coordinate_transformation_mode",
```
maybe?

daquexian replied:
Looks much better than the old one. I'll update it.

wschin commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1395,40 +1395,94 @@ ONNX_OPERATOR_SET_SCHEMA(
             "The scale array along each dimension. It takes value greater than 0. If it's less than 1,"
             " it's sampling down, otherwise, it's upsampling. The number of elements of 'scales' should"
-            " be the same as the rank of input 'X'.",
+            " be the same as the rank of input 'X'. It will be ignored if 'sizes' is specified.",
```
```suggestion
            " be the same as the rank of input 'X'. Only one of 'scales' and 'sizes' can be specified. If 'size' is needed, the user can use an empty string as the name of 'scales' in this operator's input list.",
```

wschin commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1395,40 +1395,94 @@ ONNX_OPERATOR_SET_SCHEMA(
+            3,
+            "sizes",
+            "The size of the output tensor. The number of elements of 'sizes' should be the same as the"
+            " rank of input 'X'. 'scales' will be ignored if 'sizes' is specified.",
```
```suggestion
            " rank of input 'X'. Only one of 'scales' and 'sizes' can be specified.",
```

wschin commented on onnx/backend/test/case/node/resize.py:
```diff
@@ -8,15 +8,233 @@
+                          ):                         # type: (...) -> np.ndarray
+    def get_neighbor_idxes(x, n, limit):  # type: (float, int, int) -> np.ndarray
+        """
+        Return the n nearest indexes, prefer the indexes smaller than x
```
```suggestion
        Return the n nearest indexes to x among [0, limit), prefer the indexes smaller than x.
```

wschin commented on onnx/backend/test/case/node/resize.py:
```diff
@@ -8,15 +8,233 @@
+        Pad `data` in 'edge' mode, and get n nearest elements in the padded array and their indexes in the original
+        array
+        :param x:
+        :param n:  the number of the wanted elements
```
```suggestion
        :param n: the expected element number in the padded x.
```

wschin commented on onnx/backend/test/case/node/resize.py:
```diff
@@ -8,15 +8,233 @@
+        """
+        Pad `data` in 'edge' mode, and get n nearest elements in the padded array and their indexes in the original
+        array
+        :param x:
```
```suggestion
        :param x:
```
```suggestion
        :param x: vector to be padded.
```

wschin commented on onnx/backend/test/case/node/resize.py:
```diff
@@ -8,15 +8,233 @@
+        padded = np.pad(data, pad_width, mode='edge')
+        x += pad_width
+
+        idxes = get_neighbor_idxes(x, n, len(padded))
```
What if `x > len(padded)` or `x + n/2 > len(padded)`?

daquexian replied:
it will not happen, after `x += pad_width`, x is in [n/2, len(data) - 1 + n/2], while len(padded) is len(data) + n

wschin commented on onnx/backend/test/case/node/resize.py:
```diff
@@ -8,15 +8,233 @@
+                coeffs[i] = 0
+        coeffs /= sum(coeffs)
+
+    return np.dot(coeffs, points).item()
```
Your code looks very inspiring. It basically decompose Resize into several abstractive steps:

- roi selection
- coordinate transformation
- neighbor selection
- dot product (interpolation)
It's not a blocking comment but I am wondering if we can further decompose Resize in the future.

daquexian commented:
@wschin thanks for your valuable comments. I have updated this pr accordingly

wschin commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1602,40 +1602,104 @@ ONNX_OPERATOR_SET_SCHEMA(
+
+The coordinate of each dimension is transformed individually. Denote x_resized as the coordinate of a specific dimension in the resized tensor, x_original as the coordinate of this dimension in the original tensor, length_original as the length of the original tensor in this dimension, length_resized as the length of the resized tensor in this dimension, roi_x = (start, end) of the this dimension ...
+
+if scaler is "half_pixel", <br/>
```
```suggestion
if coordinate_transformation_mode is "half_pixel", <br/>
```
and also for the following places.

daquexian replied:
Thanks! I have updated it.

wschin commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1602,40 +1602,104 @@ ONNX_OPERATOR_SET_SCHEMA(
+static const char* Resize_attr_coordinate_transformation_mode_doc = R"DOC(
+This attribute describes how to transform the coordinate in the resized tensor to the coordinate in the original tensor. <br/>
+
+The coordinate of each dimension is transformed individually. Denote x_resized as the coordinate of a specific dimension in the resized tensor, x_original as the coordinate of this dimension in the original tensor, length_original as the length of the original tensor in this dimension, length_resized as the length of the resized tensor in this dimension, roi_x = (start, end) of the this dimension ...
```
```suggestion
The coordinate of each dimension is transformed individually. Let's describe a case using axis x as an example. 
Denote x_resized as the coordinate of a specific dimension in the resized tensor, x_original as the coordinate of this dimension in the original tensor, length_original as the length of the original tensor in this dimension, length_resized as the length of the resized tensor in this dimension, roi_x = (startx, endx) of the this dimension in input "roi", scale = length_resized / length_original, <br/>
```

daquexian replied:
Great, your help is precious!

wschin commented on docs/Changelog.md:
```diff
@@ -10746,6 +10746,83 @@ This version of the operator has been available since version 11 of the default
+x_original = (x_resized + 0.5) / scale, <br/>
+
+if scaler is "tf_crop_and_resize", <br/>
+x_original = length_resized > 1 ? roi_x[0] * (length_original - 1) + x_resized * (roi_x[1] - roi_x[0]) * (length_original - 1) / (length_resized - 1) : 0.5 * (roi_x[0] + roi_x[1]) * (length_original - 1).</dd>
```
```suggestion
x_original = length_resized > 1 ? roi_x[0] * (length_original - 1) + x_resized * (endx[1] - startx[0]) * (length_original - 1) / (length_resized - 1) : 0.5 * (endx[0] + startx[1]) * (length_original - 1).</dd>
```

daquexian replied:
Updated in b067855, thanks!

wschin commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1602,40 +1602,104 @@ ONNX_OPERATOR_SET_SCHEMA(
+x_original = (x_resized + 0.5) / scale, <br/>
+
+if scaler is "tf_crop_and_resize", <br/>
+x_original = length_resized > 1 ? roi_x[0] * (length_original - 1) + x_resized * (roi_x[1] - roi_x[0]) * (length_original - 1) / (length_resized - 1) : 0.5 * (roi_x[0] + roi_x[1]) * (length_original - 1).)DOC";
```
```suggestion
x_original = length_resized > 1 ? roi_x[0] * (length_original - 1) + x_resized * (endx[1] - startx[0]) * (length_original - 1) / (length_resized - 1) : 0.5 * (endx[0] + startx[1]) * (length_original - 1).)DOC";
```

daquexian replied:
Thanks! It is now updated in b067855

wschin commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1602,40 +1602,104 @@ ONNX_OPERATOR_SET_SCHEMA(
-static const char* Resize_ver10_doc = R"DOC(
-Resize the input tensor.
+static const char* Resize_ver11_doc = R"DOC(
+Resize the input tensor. In general, it calculates every value in the output tensor as a weighted average of neighborhood (a.k.a. sampling locations) in the input tensor.
```
```suggestion
Resize the input tensor. In general, it calculates every value in the output tensor as a weighted average of neighborhood (a.k.a. sampled locations) in the input tensor.
```

daquexian replied:
"sampling locations" is from https://github.com/tensorflow/tensorflow/blob/3ae2c6691b7c6e0986d97b150c9283e5cc52c15f/tensorflow/core/kernels/resize_bicubic_op.cc#L98. Do you think "sampled locations" is natural?

wschin commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1602,40 +1602,104 @@ ONNX_OPERATOR_SET_SCHEMA(
+x_original = (x_resized + 0.5) / scale, <br/>
+
+if scaler is "tf_crop_and_resize", <br/>
+x_original = length_resized > 1 ? roi_x[0] * (length_original - 1) + x_resized * (roi_x[1] - roi_x[0]) * (length_original - 1) / (length_resized - 1) : 0.5 * (roi_x[0] + roi_x[1]) * (length_original - 1).)DOC";
```
I really like your detailed documentation. It truly standardize something.

wschin commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1602,40 +1602,104 @@ ONNX_OPERATOR_SET_SCHEMA(
             "mode",
-            "Two interpolation modes: nearest (default), and linear (including bilinear, trilinear, etc)",
+            "Three interpolation modes: nearest (default), linear and cubic. "
+            "The \"linear\" mode includes linear interpolation for 1D tensor and N-D linear interpolation for N-D tensor (For example, bilinear interpolation for 2D tensor). "
```
```suggestion
            "The \"linear\" mode includes linear interpolation for 1D tensor and N-D linear interpolation for N-D tensor (for example, bilinear interpolation for 2D tensor). "
```

wschin commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1602,40 +1602,104 @@ ONNX_OPERATOR_SET_SCHEMA(
-            "Two interpolation modes: nearest (default), and linear (including bilinear, trilinear, etc)",
+            "Three interpolation modes: nearest (default), linear and cubic. "
+            "The \"linear\" mode includes linear interpolation for 1D tensor and N-D linear interpolation for N-D tensor (For example, bilinear interpolation for 2D tensor). "
+            "The \"cubic\" mode includes cubic interpolation for 1D tensor and N-D cubic interpolation for N-D tensor (For example, bicubic interpolation for 2D tensor)."
```
```suggestion
            "The \"cubic\" mode includes cubic interpolation for 1D tensor and N-D cubic interpolation for N-D tensor (for example, bicubic interpolation for 2D tensor)."
```

wschin commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1602,40 +1602,104 @@ ONNX_OPERATOR_SET_SCHEMA(
+            AttributeProto::STRING,
+            std::string("round_prefer_floor"))
+        .Attr("extrapolation_value",
+            "When scaler is \"tf_crop_and_resize\" and x_original is outside the range [0, length_original - 1], this value is used as the corresponding output value. Default is 0.0f.",
```
```suggestion
            "When scaler is \"tf_crop_and_resize\" and x_original is outside the range [0, length_original - 1], this value is used as the corresponding output value. Default is 0.0f. This attribute is valid only if \"exclude_outside\" is 0.",
```
Is my understanding correct? If so, could you please add a check in your shape inference code? Thank you.

daquexian replied:
It is not so. "exclude_outside" is about the neighbors outside of the input tensor in the weighted average, "extrapolation_value" is about the "x_original".

wschin commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1602,40 +1602,104 @@ ONNX_OPERATOR_SET_SCHEMA(
+            AttributeProto::STRING,
+            std::string("half_pixel"))
+        .Attr("nearest_mode",
+            "Four modes: round_prefer_floor (default), round_prefer_ceil, floor, ceil. Only used by nearest interpolation. It indicates how to get \"nearest\" pixel in input tensor from x_original.",
```
```suggestion
            "Four modes: round_prefer_floor (default), round_prefer_ceil, floor, ceil. Only used by nearest interpolation. It indicates how to get \"nearest\" pixel in input tensor from x_original, so this attribute is valid only if \"mode\" is \"nearest\".",
```

wschin commented on onnx/defs/tensor/defs.cc:
```diff
@@ -1602,40 +1602,104 @@ ONNX_OPERATOR_SET_SCHEMA(
+            AttributeProto::STRING,
+            std::string("round_prefer_floor"))
+        .Attr("extrapolation_value",
+            "When scaler is \"tf_crop_and_resize\" and x_original is outside the range [0, length_original - 1], this value is used as the corresponding output value. Default is 0.0f.",
```
```suggestion
            "When \"coordinate_transform_mode\" is \"tf_crop_and_resize\" and x_original is outside the range [0, length_original - 1], this value is used as the corresponding output value. Default is 0.0f.",
```

daquexian replied:
Thanks! They are now replaced in 66d2003

lara-hdr commented:
Thank you @daquexian for the updates.
The CI is failing with the error "Unrecognized attribute: scaler for operator Resize" when checking the models, since the attribute was renamed.
Could you regenerate them to fix the issue?

daquexian commented:
@wschin @lara-hdr I have updated this PR :) Travis CI failed due to network error
